### PR TITLE
Feat/scout 23 competitions module

### PR DESCRIPTION
**Pull request by dejwi**
[SCOUT-23](https://playmakerpro.atlassian.net/browse/SCOUT-23?atlOrigin=eyJpIjoiNTY4ZDVjMWZkN2U1NGRkNmI4ZGFmODc0Y2YyZmU2ODQiLCJwIjoiaiJ9)

### Task Description
Competitions module
- table
- create/edit form
- translations
- index/create/edit/id pages
- add to panel
- edit/delete admin
- create/edit pages admin
- other components base

### Additional Notes (optional)
- can't get data from endpoint using other role than admin
- some unfinished endpoints as before, mostly marked with `// UPDATE IN FUTURE`
- I'm not sure about my translations🥲

### CODE REVIEW

File: src/modules/competition-age-categories/combot.tsx [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import { AutocompleteRenderInputParams, TextField } from '@mui/material'
```
Comment (jakubfiglak) [inline src/modules/competition-age-categories/combot.tsx:1]: There's a typo in file name, please change to `combo` :)

File: src/modules/competitions/details-card.tsx [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  CardHeader,
+  Grid
+} from '@mui/material'
+import { useTranslation } from 'next-i18next'
+
+import { CardItemBasic } from '@/components/details-card-item'
+import { CompetitionIcon } from '@/components/icons'
+import { CompetitionDto } from '@/modules/competitions/types'
+
+type IDetailsCard = {
```
Comment (jakubfiglak) [inline src/modules/competitions/details-card.tsx:14]: Please do not prefix type name with an `I`. This prefix is there to indicate this is an `interface`. I think I've used interfaces for components props everywhere, so I think we should keep consistency. Also, it should be suffixed with `Props` -> `IDetailsCardProps`

File: src/modules/competitions/details-card.tsx [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  comp: CompetitionDto
```
Comment (jakubfiglak) [inline src/modules/competitions/details-card.tsx:15]: I'd rather not use such shortcuts, please use full `competition` here

File: src/modules/competitions/forms/create.tsx [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+              helperText={touched.ageCategoryId ? errors.ageCategoryId : undefined}
+            />
+            <CountriesCombo
+              data={countriesData}
+              name='countryId'
+              label={t('COUNTRY')}
+              error={touched.countryId && !!errors.countryId}
+              helperText={touched.countryId ? errors.countryId : undefined}
+            />
+            <Field
```
Comment (jakubfiglak) [inline src/modules/competitions/forms/create.tsx:81]: We cannot let users pick negative values here or 0. Also, the value needs to be an integer. We need to think of a way to ensure that. I think the easiest solution would be to create a select field here (with values of `1, 2, 3, ..., 15`) but if you prefer handling this another way, please do.  

PS. We have the same issue in `FilterForm`

File: src/modules/competitions/forms/filter.tsx [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+          <Container>
+            <Field
+              name="name"
+              as={TextField}
+              variant="outlined"
+              fullWidth
+              label={t('NAME')}
+              size="small"
+            />
+            <Field
```
Comment (jakubfiglak) [inline src/modules/competitions/forms/filter.tsx:58]: Same issue as in `CreateForm`

File: src/modules/competitions/forms/utils.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        required: true,
+        message: t('competitions:NO_TYPE_ERROR'),
+      }),
+      gender: yup
+        .string()
+        .oneOf(['MALE', 'FEMALE'])
+        .nullable()
+        .required(t('competitions:NO_GENDER_ERROR')),
+      juniorLevelId: validateId({
+        required: true,
```
Comment (jakubfiglak) [inline src/modules/competitions/forms/utils.ts:46]: This field is not required, please remove this constraint

File: src/modules/competitions/genders-combo.tsx [hunk 7]
```diff
@@ -0,0 +1,39 @@
+import { AutocompleteRenderInputParams, TextField } from '@mui/material'
```
Comment (jakubfiglak) [inline src/modules/competitions/genders-combo.tsx:1]: I don't think we need a combo for this - there's only two possible values. Please refactor this to a regular select input (take `FootedSelect` component as an example)

File: src/modules/competitions/table/row.tsx [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+            text={t('DELETE')}
+            onClick={() => {
+              handleMenuAction(onDeleteClick)
+            }}
+            disabled={!isDeleteOptionEnabled}
+          />
+        </TableMenu>
+      </StyledTableCell>
+      <StyledTableCell>{name}</StyledTableCell>
+      <StyledTableCell>{t(`${gender}S`)}</StyledTableCell>
```
Comment (jakubfiglak) [inline src/modules/competitions/table/row.tsx:76]: Honestly, I'd just display `M` or `K` in Polish here and `M` or `F` in English

**Review by jakubfiglak — CHANGES_REQUESTED** [review]
Actually, this module is meant to be admin-only and that's why there're errors for other user roles. Sorry if the ticket description confused you. Please move navigation item to `ADMIN PANEL` & protect all pages.

We have an issue with `level` input (please see in-code comments)

Screenshot 2022-08-12 at 20 55 46

And with `competitionJuniorLevel` validation (it should not be required)

Screenshot 2022-08-12 at 20 57 57

Other than that - great work, as always 🚀

File: src/modules/competitions/forms/utils.ts [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+  CompetitionDto,
+  CreateCompetitonDto,
+  UpdateCompetitionDto,
+} from '../types'
+
+export const initialValues: CreateCompetitonDto = {
+  name: '',
+  ageCategoryId: 0,
+  countryId: 0,
+  level: 0,
```
Comment (jakubfiglak) [inline src/modules/competitions/forms/utils.ts:17]: I'd set the default level to 1 here (0 is not even a valid value)

**Review by jakubfiglak — CHANGES_REQUESTED** [review]
Two little things left to change and we're good to go:)

**Comment by jakubfiglak** [comment]
> note: edit&create both have 1-15 level validation, filter form is left without it as min/max props dont work and I think yup validation is bit too much in a filter form as it doesnt affect data, but can be added if needed

You're right - yup is an overkill in case of filter form. But, we can use native HTML validation and set `inputProps` on our field to set `min`, `max`, `step` & perhaps `pattern` value to ensure client is not sending invalid requests to the backend.

https://stackoverflow.com/questions/8808590/number-input-type-that-takes-only-integers
https://medium.com/codex/making-html-5-numeric-inputs-only-accept-integers-d3d117973d56

**Comment by dejwi** [comment]
min/max/step/pattern only prevents invalid input when using arrows to increment value, not when typing in value so I've added little helper text as well is that okay?

image

**Comment by jakubfiglak** [comment]
> min/max/step/pattern only prevents invalid input when using arrows to increment value, not when typing in value so I've added little helper text as well is that okay?
> 
> image

That's not exactly true - native html validation will prevent user from submitting the form if any of the values doesn't pass the validation (see attached screenshot)

Screenshot 2022-08-16 at 20 54 42

However the helper text thing is actually a good idea but I'd prefer using actual `helperText` prop for this. Example:

Screenshot 2022-08-16 at 20 56 29

I will approve this PR though, we don't have to fix that now I suppose

**Review by jakubfiglak — APPROVED** [review]

**Pull request merged**